adapters.go

```diff
@@ -77,32 +77,35 @@ func (a *whisperAdapter) RetrievePublicMessages(chatID string) ([]*protocol.Mess
 		return nil, err
 	}
 
-	logger := a.logger.With(zap.String("site", "RetrievePublicMessages"))
-
-	decodedMessages := make([]*protocol.Message, 0, len(messages))
-	for _, item := range messages {
-		shhMessage := whisper.ToWhisperMessage(item)
-
-		hlogger := logger.With(zap.Binary("hash", shhMessage.Hash))
-
-		hlogger.Debug("received a public message")
+	return a.handleRetrievedMessages(messages, true)
+}
 
-		statusMessage, err := a.decodeMessage(shhMessage)
-		if err != nil {
-			hlogger.Error("failed to decode message", zap.Error(err))
-			continue
-		}
+// RetrieveAllMessages retrieves the collected messages.
+// It implies joining a chat if it has not been joined yet.
+func (a *whisperAdapter) RetrieveAllMessages() ([]*protocol.ChatMessages, error) {
+	messagesByChat, err := a.transport.RetrieveAllMessages()
+	if err != nil {
+		return nil, err
+	}
 
-		switch m := statusMessage.Message.(type) {
-		case protocol.Message:
-			m.ID = statusMessage.ID
-			m.SigPubKey = statusMessage.SigPubKey
-			decodedMessages = append(decodedMessages, &m)
-		default:
-			hlogger.Error("skipped a public message of unsupported type")
+	var latest []*protocol.ChatMessages
+	for chat, messages := range messagesByChat {
+		decodedMessages, err := a.handleRetrievedMessages(messages, chat.IsPublic())
+		if err == nil { // TODO: What to do when we get an error here? Should it be best effort, or can we safely abort?
```

Comment:
Just log it. We shouldn't stop if there is an invalid message as maybe one can try to kill all clients using a corrupted messages or something.